File: sdk/dart/persona_agent_flutter/lib/src/utils/tool_grouping.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:persona_agent_client/persona_agent_client.dart';

/// Which semantic family a tool belongs to, for a cluster header — ported
/// from `sdk/ui/src/utils/toolGrouping.ts`.
String toolGroupKey(PersonaToolCall tool) {
  final name = tool.toolName.toLowerCase();
  if (name.contains('memory') || name.contains('preference')) return 'memory';
  Map<String, dynamic> args = const {};
  if (tool.args != null) {
    try {
      final decoded = jsonDecode(tool.args!);
      if (decoded is Map<String, dynamic>) args = decoded;
    } catch (_) {}
  }
  final path = args['file_path'] ?? args['path'] ?? args['filePath'];
  if (path is String && path.contains('/memories')) return 'memory';
  if (name.contains('file') || name == 'ls' || name == 'glob' || name == 'grep') return 'file';
  if (name.contains('search') || name.startsWith('tavily')) return 'search';
  if (name == 'task') return 'task';
  if (name.contains('todo')) return 'plan';
  return name;
}

class PersonaToolClusterMeta {
  const PersonaToolClusterMeta({required this.title, required this.icon});
  final String title;
  final IconData icon;
}

const Map<String, PersonaToolClusterMeta> defaultClusterLabels = {
  'memory': PersonaToolClusterMeta(title: 'Personalizing memory', icon: Icons.psychology_outlined),
  'file': PersonaToolClusterMeta(title: 'Working with files', icon: Icons.description_outlined),
  'search': PersonaToolClusterMeta(title: 'Searching the web', icon: Icons.public),
  'task': PersonaToolClusterMeta(title: 'Running subagents', icon: Icons.smart_toy_outlined),
  'plan': PersonaToolClusterMeta(title: 'Updating the plan', icon: Icons.checklist_rounded),
  'mixed': PersonaToolClusterMeta(title: 'Performing actions', icon: Icons.build_outlined),
};

PersonaToolClusterMeta clusterMeta(List<PersonaToolCall> tools, {Map<String, PersonaToolClusterMeta>? labels}) {
  final merged = {...defaultClusterLabels, ...?labels};
  final groups = tools.map(toolGroupKey).toSet();
  final key = groups.length == 1 ? groups.first : 'mixed';
  return merged[key] ?? merged['mixed']!;
}

sealed class PersonaToolGroupItem {
  const PersonaToolGroupItem();
}

class PersonaToolGroupSingle extends PersonaToolGroupItem {
  const PersonaToolGroupSingle(this.tool);
  final PersonaToolCall tool;
}

class PersonaToolGroupCluster extends PersonaToolGroupItem {
  const PersonaToolGroupCluster(this.tools);
  final List<PersonaToolCall> tools;
}

/// Clusters consecutive tool calls from one message into groups a
/// `PersonaToolGroup` can render as one collapsible unit — `present_file`
/// never joins a group, since its whole purpose is "highlight this file".
List<PersonaToolGroupItem> groupToolCalls(List<PersonaToolCall> toolCalls) {
  final items = <PersonaToolGroupItem>[];
  var buffer = <PersonaToolCall>[];

  void flush() {
    if (buffer.isEmpty) return;
    items.add(buffer.length == 1 ? PersonaToolGroupSingle(buffer.single) : PersonaToolGroupCluster(buffer));
    buffer = [];
  }

  for (final tool in toolCalls) {
    if (tool.toolName == 'present_file') {
      flush();
      items.add(PersonaToolGroupSingle(tool));
    } else {
      buffer.add(tool);
    }
  }
  flush();
  return items;
}

File: sdk/ui/src/utils/toolGrouping.ts
import type { ComponentType } from 'react';
import type { PersonaToolCall } from '@personaai/react';

export interface PersonaToolClusterMeta {
  title: string;
  icon?: ComponentType<{ className?: string }>;
}

/** Keyed by `toolGroupKey`'s return value — `mixed` is the fallback when a group's tools don't share one key. */
export type PersonaToolClusterLabels = Record<string, PersonaToolClusterMeta>;

function safeParseArgs(args: string | undefined): Record<string, unknown> {
  if (!args) return {};
  try {
    const parsed = JSON.parse(args);
    return typeof parsed === 'object' && parsed ? parsed : {};
  } catch {
    return {};
  }
}

/**
 * Which semantic family a tool belongs to, for the cluster header. Memory is
 * detected by name AND by file ops touching /memories/ paths, matching how
 * persona.hasanraiyan.me's own frontend classifies tools for the same purpose.
 */
export function toolGroupKey(tool: PersonaToolCall): string {
  const name = tool.toolName.toLowerCase();
  if (name.includes('memory') || name.includes('preference')) return 'memory';
  const args = safeParseArgs(tool.args);
  const path = (args.file_path ?? args.path ?? args.filePath ?? '') as unknown;
  if (typeof path === 'string' && path.includes('/memories')) return 'memory';
  if (name.includes('file') || name === 'ls' || name === 'glob' || name === 'grep') return 'file';
  if (name.includes('search') || name.startsWith('tavily')) return 'search';
  if (name === 'task') return 'task';
  if (name.includes('todo')) return 'plan';
  return name;
}

export interface PersonaToolGroupItem {
  type: 'single' | 'group';
  tools: PersonaToolCall[];
}

/**
 * Clusters consecutive tool calls from one message into groups a
 * `PersonaToolGroup` can render as one collapsible unit, instead of one card
 * per call. `present_file` never joins a group — its whole purpose is
 * "highlight this file", which a generic "N steps" cluster header would bury.
 */
export function groupToolCalls(toolCalls: PersonaToolCall[]): PersonaToolGroupItem[] {
  const items: PersonaToolGroupItem[] = [];
  let buffer: PersonaToolCall[] = [];

  const flush = () => {
    if (buffer.length === 0) return;
    items.push({ type: buffer.length === 1 ? 'single' : 'group', tools: buffer });
    buffer = [];
  };

  for (const tool of toolCalls) {
    if (tool.toolName === 'present_file') {
      flush();
      items.push({ type: 'single', tools: [tool] });
    } else {
      buffer.push(tool);
    }
  }
  flush();
  return items;
}
